# Lecture 3.2:  Data transformation

<div style="border: 1px double black; padding: 10px; margin: 10px">

**Goals for today's lecture:**
* Continue to learn [how to manipulate data](#Data-manipulation), including:
    * Pipes
    * Adding New Variables
    
* We will go through the data set `flight` 
    
We will answer question such as:     
* What days of the year / week are the busiest for flying?    
    
This lecture note corresponds to Chapter 5.5 of your book.
    
    
</div>


Let us load up the `tidyverse` and `nycflights13` packages.



We will start with the `flight` data set that we use in the previous lecture. 

In [2]:
library(tidyverse)
library(nycflights13)
head(flights)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──

✔ ggplot2 3.3.5     ✔ purrr   0.3.4
✔ tibble  3.1.3     ✔ dplyr   1.0.7
✔ tidyr   1.1.3     ✔ stringr 1.4.0
✔ readr   2.0.1     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time
1 2013 1     1   517      515             2         830      819          
2 2013 1     1   533      529             4         850      830          
3 2013 1     1   542      540             2         923      850          
4 2013 1     1   544      545            -1        1004     1022          
5 2013 1     1   554      600            -6         812      837          
6 2013 1     1   554      558            -4         740      728          
  arr_delay carrier flight tailnum origin dest air_time distance hour minute
1  11       UA      1545   N14228  EWR    IAH  227      1400     5    15    
2  20       UA      1714   N24211  LGA    IAH  227      1416     5    29    
3  33       AA      1141   N619AA  JFK    MIA  160      1089     5    40    
4 -18       B6       725   N804JB  JFK    BQN  183      1576     5    45    
5 -25       DL       461   N668DN  LGA    ATL  116       762     6     0    
6  12       UA      1696   N39463  EWR    ORD  150       719     5    58    
  time_hour          
1 2013-01-01 05:00:00
2 2013-01-01 05:00:00
3 2013-01-01 05:00:00
4 2013-01-01 05:00:00
5 2013-01-01 06:00:00
6 2013-01-01 05:00:00

# Pipes
Starting now, we will make extensive use of the pipe operator `%>%`. 

### How `%>%` works
Under the hood, `x %>% f(y)` turns into `f(x, y)`, and `x %>% f(y) %>% g(z)` turns into `g(f(x, y), z)` and so on. We can use `%>%` on any function, not just those defined in tidyverse.

Here is an example on printing "hello world" using pipe. 

In [12]:
"hello world" %>% print()  # prints "hello world"

[1] "hello world"


We will see the usefulness of pipe `%>%` later in the lecture as it greatly simplifies our code

# Adding New Variables
The `dplyr`/`tidyverse` package offers the `mutate()` and `transmute()` commands to add new variables to data tibbles. The syntax is:
```{r}
<tibble> %>% mutate(<new variable> = <formula for new variable>,
                    <other new variable> = <other formula>)
```
This returns a copy of `<tibble>` with the new variables added on `transmute()` does the same thing as `mutate()` but only keeps the new variables.

Let us zoom in on a few variables of interest.

In [3]:
my_flights <- select(flights, year:day, dep_time, arr_time, air_time, origin, dest)
head(my_flights)

year month day dep_time arr_time air_time origin dest
1 2013 1     1   517       830     227      EWR    IAH 
2 2013 1     1   533       850     227      LGA    IAH 
3 2013 1     1   542       923     160      JFK    MIA 
4 2013 1     1   544      1004     183      JFK    BQN 
5 2013 1     1   554       812     116      LGA    ATL 
6 2013 1     1   554       740     150      EWR    ORD

Use Pipe `%>%` to create the table above. 

In [4]:
flights %>% select(year:day, dep_time, arr_time, air_time, origin, dest) %>% head()

year month day dep_time arr_time air_time origin dest
1 2013 1     1   517       830     227      EWR    IAH 
2 2013 1     1   533       850     227      LGA    IAH 
3 2013 1     1   542       923     160      JFK    MIA 
4 2013 1     1   544      1004     183      JFK    BQN 
5 2013 1     1   554       812     116      LGA    ATL 
6 2013 1     1   554       740     150      EWR    ORD

The above code basically select the variables that we are interested in and save it into the object `my_flights`.  

Additional variable can be added using the `mutate()` function. We already have an `air_time` variable. Let us compute the total time for the flight by subtracting the time of departure `dep_time` from time of arrival `arr_time`.

In [5]:
mutate(my_flights, total_time = arr_time - dep_time) %>%
    head()

year month day dep_time arr_time air_time origin dest total_time
1 2013 1     1   517       830     227      EWR    IAH  313       
2 2013 1     1   533       850     227      LGA    IAH  317       
3 2013 1     1   542       923     160      JFK    MIA  381       
4 2013 1     1   544      1004     183      JFK    BQN  460       
5 2013 1     1   554       812     116      LGA    ATL  258       
6 2013 1     1   554       740     150      EWR    ORD  186

Another way to do the same thing is by using pipe twice

In [6]:
flights %>% mutate(total_time = arr_time - dep_time) %>% 
            select(year, month, day, dep_time, arr_time, air_time, origin, dest, total_time) %>% head()

year month day dep_time arr_time air_time origin dest total_time
1 2013 1     1   517       830     227      EWR    IAH  313       
2 2013 1     1   533       850     227      LGA    IAH  317       
3 2013 1     1   542       923     160      JFK    MIA  381       
4 2013 1     1   544      1004     183      JFK    BQN  460       
5 2013 1     1   554       812     116      LGA    ATL  258       
6 2013 1     1   554       740     150      EWR    ORD  186

We notice something odd though. When we subtract 5h 17m from 8h 30m we should get 3h 13m, i.e. 193 minutes. But instead we get 313 minutes below.

The issue is that `dep_time` and `arr_time` are in the hour-minute notation, so you cannot add and subtract them like regular numbers. We should first convert these times into the number of minutes elapsed since midnight.

We want add to new variables `new_dep` and `new_arr` but we need to write a function first that can do the conversion. The function is given below; we'll learn how it works later in the semester. For now just think of it as a black box that converts times from one format to another.

In [7]:
hourmin2min <- function(hourmin) {
    min <- hourmin %% 100 # quotient after division by 100
    hour <- (hourmin - min) %/% 100 # remainder after division by 100
    return(60*hour + min)
} 

Let us test the function on 530. That's 5h 30min, i.e., 330 minutes since midnight.

In [8]:
hourmin2min(530)

[1] 330

The `hourmin2min` function is **vectorized**: given a vector, it outputs a vector.

In [9]:
hourmin2min(c(430,530,630,730))

[1] 270 330 390 450

Let us now create two new variables obtained from `arr_time` and `dep_time` by converting them into minutes since midnight. In the same command, we can also create a new `total_time` column containing their difference.

In [10]:
my_flights_new <- mutate(my_flights, new_arr = hourmin2min(arr_time), new_dep = hourmin2min(dep_time))
head(my_flights_new)

year month day dep_time arr_time air_time origin dest new_arr new_dep
1 2013 1     1   517       830     227      EWR    IAH  510     317    
2 2013 1     1   533       850     227      LGA    IAH  530     333    
3 2013 1     1   542       923     160      JFK    MIA  563     342    
4 2013 1     1   544      1004     183      JFK    BQN  604     344    
5 2013 1     1   554       812     116      LGA    ATL  492     354    
6 2013 1     1   554       740     150      EWR    ORD  460     354

Now we can subtract the departure time `new_dep` from the arrival time `new_arr` to get a new variable `total_time`.

In [11]:
my_flights_total <- mutate(my_flights_new, total_time = new_arr - new_dep)
head(my_flights_total)

year month day dep_time arr_time air_time origin dest new_arr new_dep
1 2013 1     1   517       830     227      EWR    IAH  510     317    
2 2013 1     1   533       850     227      LGA    IAH  530     333    
3 2013 1     1   542       923     160      JFK    MIA  563     342    
4 2013 1     1   544      1004     183      JFK    BQN  604     344    
5 2013 1     1   554       812     116      LGA    ATL  492     354    
6 2013 1     1   554       740     150      EWR    ORD  460     354    
  total_time
1 193       
2 197       
3 221       
4 260       
5 138       
6 106

Let us try to do the same thing using pipe just using one line of code.  In this code, we are only interested in the following variables -- `dep_time`, `arr_time`, `new_dep`, `new_arr`, and `total_time`.  

In [12]:
 mutate(flights, new_arr = hourmin2min(arr_time), new_dep = hourmin2min(dep_time),total_time = new_arr - new_dep
) %>% select(dep_time, arr_time, new_dep, new_arr, total_time) %>% head()

dep_time arr_time new_dep new_arr total_time
1 517       830     317     510     193       
2 533       850     333     530     197       
3 542       923     342     563     221       
4 544      1004     344     604     260       
5 554       812     354     492     138       
6 554       740     354     460     106

## Up Next - Summarize Function and Case Study on Data Manipulation

Let's now review some tools that we have learnt.  Specifically, we have learnt the function:
* `filter`
* `arrange`
* `select` 

### Exercise

<table class="table-condensed">
    <tr><th>Beginner</th><th>Advanced</th></tr>
    <tr><td>How many flights were there in 
        months beginning with the letter <code>J</code>?</td>
        <td>How many flights departed on a Monday?</td>
    </tr>
<tr><td>

1. 27,004
2. 57,668
3. 84,672
4. 93,101

</td><td>

1. 46,537
2. 51,812
3. 80,100
4. 101,991

</td>
    </tr></table>

In [11]:
nrow(filter(flights, month %in% which(substr(month.name,0,1) == "J")))



[1] 84672

### Exercise

<table class="table-condensed">
    <tr><th>Beginner</th><th>Advanced</th></tr>
    <tr><td>
What proportion of the flights have a missing departure time?
        </td>
        <td>
Of all the flights that departed in the first week of January, how many have a missing departure time?
        </td>
    </tr>
<tr><td>

1. 0.003
2. 0.025
3. 0.081
4. 0.105


</td><td>

1. None
2. 35
3. 101
4. 6,064

</td>
    </tr></table>
    

In [3]:
#  Beginner question
print(filter(flights, is.na(dep_time)))
8255 / nrow(flights)

# A tibble: 8,255 x 19
    year month   day dep_time sched_dep_time dep_delay arr_time sched_arr_time
   <int> <int> <int>    <int>          <int>     <dbl>    <int>          <int>
 1  2013     1     1       NA           1630        NA       NA           1815
 2  2013     1     1       NA           1935        NA       NA           2240
 3  2013     1     1       NA           1500        NA       NA           1825
 4  2013     1     1       NA            600        NA       NA            901
 5  2013     1     2       NA           1540        NA       NA           1747
 6  2013     1     2       NA           1620        NA       NA           1746
 7  2013     1     2       NA           1355        NA       NA           1459
 8  2013     1     2       NA           1420        NA       NA           1644
 9  2013     1     2       NA           1321        NA       NA           1536
10  2013     1     2       NA           1545        NA       NA           1910
# … with 8,245 more rows, and

[1] 0.02451184

In [4]:
# Advanced question
nrow(filter(flights, is.na(dep_time), month==1, between(day,1,7)))


[1] 35

### Exercise

<table class="table-condensed">
    <tr><th>Beginner</th><th>Advanced</th></tr>
    <tr><td>
        What time did the first flight depart on the last day in February?
<td>
 
 Of all the flights that departed on or ahead of schedule in the first 15 days of any month, which one was in the air for the *second* shortest amount of time?

</td>
    </tr>
<tr><td>

1. 4:15am
2. 4:57am
3. 5:01am
4. 5:40am

</td><td>

1. EV 4118
2. HA 51
3. EV 4631
4. EV 4619

</td>
    </tr></table>

In [5]:
# Beginner
answer <- arrange(filter(flights, month == 2),
      desc(day),
      dep_time)
print(slice(answer, 1:5))

# A tibble: 5 x 19
   year month   day dep_time sched_dep_time dep_delay arr_time sched_arr_time
  <int> <int> <int>    <int>          <int>     <dbl>    <int>          <int>
1  2013     2    28      457            500        -3      639            648
2  2013     2    28      458            501        -3      748            800
3  2013     2    28      522            530        -8      832            831
4  2013     2    28      539            540        -1      836            850
5  2013     2    28      540            545        -5     1015           1023
# … with 11 more variables: arr_delay <dbl>, carrier <chr>, flight <int>,
#   tailnum <chr>, origin <chr>, dest <chr>, air_time <dbl>, distance <dbl>,
#   hour <dbl>, minute <dbl>, time_hour <dttm>


### Exercise

<table class="table-condensed">
    <tr><th>Beginner</th><th>Advanced</th></tr>
    <tr><td>
        How many columns in flights contain the word time?
<td>
    
How many column names in flights do *not* contain `s`?

</td>
    </tr>
<tr>
    </tr></table>

In [6]:
# Beginner
ncol(select(flights, contains("time")))
select(flights, contains("time"))

[1] 6

dep_time sched_dep_time arr_time sched_arr_time air_time
1      517      515             830      819           227     
2      533      529             850      830           227     
3      542      540             923      850           160     
4      544      545            1004     1022           183     
5      554      600             812      837           116     
6      554      558             740      728           150     
7      555      600             913      854           158     
8      557      600             709      723            53     
9      557      600             838      846           140     
10     558      600             753      745           138     
11     558      600             849      851           149     
12     558      600             853      856           158     
13     558      600             924      917           345     
14     558      600             923      937           361     
15     559      600             941      910           257     
16     559      559             702      706            44     
17     559      600             854      902           337     
18     600      600             851      858           152     
19     600      600             837      825           134     
20     601      600             844      850           147     
21     602      610             812      820           170     
22     602      605             821      805           105     
23     606      610             858      910           152     
24     606      610             837      845           128     
25     607      607             858      915           157     
26     608      600             807      735           139     
27     611      600             945      931           366     
28     613      610             925      921           175     
29     615      615            1039     1100           182     
30     615      615             833      842           120     
⋮      ⋮        ⋮              ⋮        ⋮              ⋮       
336747 2123     2125           2223     2247            45     
336748 2127     2129           2314     2323            72     
336749 2128     2130           2328     2359           213     
336750 2129     2059           2230     2232            45     
336751 2131     2140           2225     2255            36     
336752 2140     2140             10       40           298     
336753 2142     2129           2250     2239            47     
336754 2145     2145            115      140           192     
336755 2147     2137             30       27           139     
336756 2149     2156           2245     2308            37     
336757 2150     2159           2250     2306            39     
336758 2159     1845           2344     2030            50     
336759 2203     2205           2339     2331            61     
336760 2207     2140           2257     2250            97     
336761 2211     2059           2339     2242           120     
336762 2231     2245           2335     2356            48     
336763 2233     2113            112       30           318     
336764 2235     2001             59     2249           123     
336765 2237     2245           2345     2353            43     
336766 2240     2245           2334     2351            41     
336767 2240     2250           2347        7            52     
336768 2241     2246           2345        1            47     
336769 2307     2255           2359     2358            33     
336770 2349     2359            325      350           196     
336771   NA     1842             NA     2019            NA     
336772   NA     1455             NA     1634            NA     
336773   NA     2200             NA     2312            NA     
336774   NA     1210             NA     1330            NA     
336775   NA     1159             NA     1344            NA     
336776   NA      840             NA     1020            NA     
       time_hour          
1      2013-

In [7]:
# Advanced
ncol(flights)-ncol(select(flights, contains("s")))

[1] 15

## Question:  What days of the year / at what airport are the busiest for flying?

Let's think about the table we would want to have in order to answer this question. Ideally,
it would look something like this:

    # A tibble: 1,095 x 4
       month   day airport n_sched_departures
       <int> <int> <chr>                <int>
     1     1     1 EWR                    305
     2     1     1 JFK                    297
     3     1     1 LGA                    240
     4     1     2 EWR                    350
     5     1     2 JFK                    321
     6     1     2 LGA                    272
     7     1     3 EWR                    336
     8     1     3 JFK                    318
     9     1     3 LGA                    260
    10     1     4 EWR                    339
    # … with 1,085 more rows

The table we are given has ~337k rows, one for each flight. How do we go from the `flights` table to the one shown above?

## Summaries
`summarize()` can be used to summarize entire data frames by collapsing them into single number summaries. The syntax is:
```{r}
summarize(<grouped tibble>, 
<new variable> = 
<formula for new variable>,
<other new variable> = <other formula>)
```

The most basic use of summarize is to compute statistics over the whole data set. 

Let us calculate the average time for departure delay `dep_delay`. 

In [8]:
tbl <- filter(flights, !is.na(dep_delay))
print(summarize(tbl, delay = mean(dep_delay)))

# A tibble: 1 x 1
  delay
  <dbl>
1  12.6


## Grouping observations
`summarize()` is most useful when combined with `group_by()` to group observations before calculating the summary statistic. The `group_by` function tells `R` how your data are grouped:

In [9]:
?group_by
tbl <- group_by(flights, month)
# print(tbl)
tbl1 <- filter(tbl, !is.na(dep_delay))
print(summarize(tbl1, delay = mean(dep_delay), 
                total_delay = sum(dep_delay)))

`summarise()` ungrouping output (override with `.groups` argument)



# A tibble: 12 x 3
   month delay total_delay
   <int> <dbl>       <dbl>
 1     1 10.0       265801
 2     2 10.8       256251
 3     3 13.2       370001
 4     4 13.9       385554
 5     5 13.0       366658
 6     6 20.8       567729
 7     7 21.7       618916
 8     8 12.6       363715
 9     9  6.72      182327
10    10  6.24      178909
11    11  5.44      146945
12    12 16.6       449394


`summarize()` applies the summary function to each group of data. Remember that it always returns **one row per group**.

In [10]:
print(summarize(group_by(flights, month), mean_dep_delay = mean(dep_delay, na.rm=T)))

`summarise()` ungrouping output (override with `.groups` argument)



# A tibble: 12 x 2
   month mean_dep_delay
   <int>          <dbl>
 1     1          10.0 
 2     2          10.8 
 3     3          13.2 
 4     4          13.9 
 5     5          13.0 
 6     6          20.8 
 7     7          21.7 
 8     8          12.6 
 9     9           6.72
10    10           6.24
11    11           5.44
12    12          16.6 


It's as if `summarize()` filtered your data for each group, calculated the summary statistic, and
then combined all the results back into one table.

In [11]:
df <- filter(flights, month == 3)
mean(df$dep_delay, na.rm = T)  

[1] 13.22708

Many summary functions are available:

* Center: `mean(), median()`
* Spread: `sd(), IQR(), mad()`
* Range: `min(), max(), quantile()`
* Position: `first(), last(), nth()`
* Count: `n(), n_distinct()`
* Logical: `any(), all()`

### Examples
The `n()` function calculates the number of rows in each group:

Let us write some code to output the number of flights for each month.   Then output the number of flights in December.   

In [12]:
tbl <- group_by(flights, month)
print(summarize(tbl, n = n()))
nrow(filter(flights, month == 12))

`summarise()` ungrouping output (override with `.groups` argument)



# A tibble: 12 x 2
   month     n
   <int> <int>
 1     1 27004
 2     2 24951
 3     3 28834
 4     4 28330
 5     5 28796
 6     6 28243
 7     7 29425
 8     8 29327
 9     9 27574
10    10 28889
11    11 27268
12    12 28135


[1] 28135

Now we are ready to generate the following table using the tools that we have learnt

    # A tibble: 1,095 x 4
       month   day airport n_sched_departures
       <int> <int> <chr>                <int>
     1     1     1 EWR                    305
     2     1     1 JFK                    297
     3     1     1 LGA                    240
     4     1     2 EWR                    350
     5     1     2 JFK                    321
     6     1     2 LGA                    272
     7     1     3 EWR                    336
     8     1     3 JFK                    318
     9     1     3 LGA                    260
    10     1     4 EWR                    339
    # … with 1,085 more rows

Use this table to answer the question: which day of the year is busiest, and at what airport?

In [13]:
tbl <- group_by(flights, month, day, origin) 
tbl1 <- summarize(tbl, n_sched_dep = n())
tbl2 <- arrange(tbl1, desc(n_sched_dep))
tbl3 <- group_by(tbl1, origin)
print(tbl1)
print(tbl2)
print(tbl3)
top_n(tbl3, 3, n_sched_dep)

`summarise()` regrouping output by 'month', 'day' (override with `.groups` argument)



# A tibble: 1,095 x 4
# Groups:   month, day [365]
   month   day origin n_sched_dep
   <int> <int> <chr>        <int>
 1     1     1 EWR            305
 2     1     1 JFK            297
 3     1     1 LGA            240
 4     1     2 EWR            350
 5     1     2 JFK            321
 6     1     2 LGA            272
 7     1     3 EWR            336
 8     1     3 JFK            318
 9     1     3 LGA            260
10     1     4 EWR            339
# … with 1,085 more rows
# A tibble: 1,095 x 4
# Groups:   month, day [365]
   month   day origin n_sched_dep
   <int> <int> <chr>        <int>
 1     4    15 EWR            377
 2     4    11 EWR            376
 3     4    18 EWR            376
 4     4    10 EWR            375
 5     4    17 EWR            375
 6     4    12 EWR            374
 7     4    19 EWR            374
 8     4    22 EWR            372
 9     4    25 EWR            372
10     4    29 EWR            372
# … with 1,085 more rows
# A tibble: 1,095 x 4
# Groups: 

month day origin n_sched_dep
1  4     11  EWR    376        
2  4     15  EWR    377        
3  4     18  EWR    376        
4  7     10  JFK    331        
5  7     11  JFK    332        
6  7     12  JFK    331        
7  9      9  LGA    345        
8  9     12  LGA    345        
9  9     13  LGA    346        
10 9     16  LGA    345        
11 9     23  LGA    345

#### A shortcut
`summarize(object =  n())` occurs so often that there is a shortcut for it: and the function is called `count` 

Let us try to output the number of flights for each month again.  Previously we use the following code to do it:

In [14]:
tbl <- group_by(flights, month)
print(summarize(tbl, n = n()))
flights %>% group_by(month) %>% summarize(n = n()) %>% print()

`summarise()` ungrouping output (override with `.groups` argument)



# A tibble: 12 x 2
   month     n
   <int> <int>
 1     1 27004
 2     2 24951
 3     3 28834
 4     4 28330
 5     5 28796
 6     6 28243
 7     7 29425
 8     8 29327
 9     9 27574
10    10 28889
11    11 27268
12    12 28135


`summarise()` ungrouping output (override with `.groups` argument)



# A tibble: 12 x 2
   month     n
   <int> <int>
 1     1 27004
 2     2 24951
 3     3 28834
 4     4 28330
 5     5 28796
 6     6 28243
 7     7 29425
 8     8 29327
 9     9 27574
10    10 28889
11    11 27268
12    12 28135


### Question -- output the number of flights for each carrier

In [15]:
top_n(summarize(group_by(flights, carrier), n = n()), 5)

`summarise()` ungrouping output (override with `.groups` argument)

Selecting by n



carrier n    
1 AA      32729
2 B6      54635
3 DL      48110
4 EV      54173
5 UA      58665

### Exercise

Use `summarize()`, `count()`, `filter()`, `arrange()` and/or `top_n()` to answer:

<table class="table-condensed">
    <tr><th>Beginner</th><th>Advanced</th></tr>
    <tr><td>
        Which plane (tail number) flew the most flights in July?
        </td>
    <td>
        How many planes flew at least one flight in January, but none in February?
        </td>
    </tr>
    <tr>
<td>
</tr>
</table>

In [16]:
# Beginner
tbl0 <- filter(flights, month == 7, !is.na(tailnum))
#print(tbl0)

tbl1 <- count(tbl0, tailnum)
print(tbl1)

top_n(tbl1,1)



# A tibble: 3,215 x 2
   tailnum     n
   <chr>   <int>
 1 D942DN      1
 2 N0EGMQ     11
 3 N10156      7
 4 N102UW      6
 5 N103US      6
 6 N104UW      8
 7 N10575     17
 8 N105UW      5
 9 N107US      9
10 N108UW      7
# … with 3,205 more rows


Selecting by n



tailnum n 
1 N298JB  76